analysis/src/inference.rs

```diff
@@ -0,0 +1,176 @@
+use ast::asm_analysis::{AnalysisASMFile, Expression, FunctionStatement, Machine};
+use number::FieldElement;
+
+pub fn infer<T: FieldElement>(file: AnalysisASMFile<T>) -> Result<AnalysisASMFile<T>, Vec<String>> {
+    let mut errors = vec![];
+    let mut res = AnalysisASMFile::default();
+
+    for (name, m) in file.machines {
+        match infer_machine(m) {
+            Ok(m) => {
+                res.machines.insert(name, m);
+            }
+            Err(e) => {
+                errors.extend(e);
+            }
+        }
+    }
+
+    if !errors.is_empty() {
+        Err(errors)
+    } else {
+        Ok(res)
+    }
+}
+
+fn infer_machine<T: FieldElement>(mut machine: Machine<T>) -> Result<Machine<T>, Vec<String>> {
+    let mut errors = vec![];
+
+    for f in machine.functions.iter_mut() {
+        for s in f.body.statements.iter_mut() {
+            if let FunctionStatement::Assignment(a) = s {
+                let expr_reg = match &*a.rhs {
+                    Expression::FunctionCall(c) => {
+                        let instr = machine
+                            .instructions
+                            .iter()
+                            .find(|i| i.name == c.id)
+                            .unwrap();
+                        let output = {
+                            let outputs = instr.params.outputs.as_ref().unwrap();
+                            assert!(outputs.params.len() == 1);
+                            &outputs.params[0]
+                        };
```

Comment:
Maybe we should have helpers for these somewhere. Matching on a FunctionCall sounds like something that's likely done in other places too.